In [4]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [5]:
Questions=r'Questions.csv'
Tags=r'Tags.csv'

In [6]:
tags=pd.read_csv(Tags,encoding='latin 1')
questions=pd.read_csv(Questions,encoding='latin 1')

In [7]:
questions.head()

Id  OwnerUserId          CreationDate            ClosedDate  Score  \
0   80         26.0  2008-08-01T13:57:07Z                   NaN     26   
1   90         58.0  2008-08-01T14:41:24Z  2012-12-26T03:45:49Z    144   
2  120         83.0  2008-08-01T15:50:08Z                   NaN     21   
3  180    2089740.0  2008-08-01T18:42:19Z                   NaN     53   
4  260         91.0  2008-08-01T23:22:08Z                   NaN     49   

                                               Title  \
0  SQLStatement.execute() - multiple queries in o...   
1  Good branching and merging tutorials for Torto...   
2                                  ASP.NET Site Maps   
3                 Function for creating color wheels   
4  Adding scripting functionality to .NET applica...   

                                                Body  
0  <p>I've written a database generation script i...  
1  <p>Are there any really good tutorials explain...  
2  <p>Has anyone got experience creating <strong>...  
3  <p>This is something I've pseudo-solved many t...  
4  <p>I have a little game written in C#. It uses...

In [8]:
tags.head()

Id             Tag
0  80            flex
1  80  actionscript-3
2  80             air
3  90             svn
4  90     tortoisesvn

In [9]:
print('Shape of Questions dataset:', questions.shape)
print('Shape of Tags dataset:', tags.shape)

Shape of Questions dataset: (1264216, 7)
Shape of Tags dataset: (3750994, 2)


In [10]:
print('Number of unique Scores:',questions['Score'].nunique())
questions[questions.Score>=5].shape

Number of unique Scores: 532


(93153, 7)

In [11]:
questions=questions[questions.Score>=5]
print(questions.shape)
questions.head()

(93153, 7)


Id  OwnerUserId          CreationDate            ClosedDate  Score  \
0   80         26.0  2008-08-01T13:57:07Z                   NaN     26   
1   90         58.0  2008-08-01T14:41:24Z  2012-12-26T03:45:49Z    144   
2  120         83.0  2008-08-01T15:50:08Z                   NaN     21   
3  180    2089740.0  2008-08-01T18:42:19Z                   NaN     53   
4  260         91.0  2008-08-01T23:22:08Z                   NaN     49   

                                               Title  \
0  SQLStatement.execute() - multiple queries in o...   
1  Good branching and merging tutorials for Torto...   
2                                  ASP.NET Site Maps   
3                 Function for creating color wheels   
4  Adding scripting functionality to .NET applica...   

                                                Body  
0  <p>I've written a database generation script i...  
1  <p>Are there any really good tutorials explain...  
2  <p>Has anyone got experience creating <strong>...  
3  <p>This is something I've pseudo-solved many t...  
4  <p>I have a little game written in C#. It uses...

In [12]:
questions.drop(columns=['OwnerUserId', 'CreationDate', 'ClosedDate', 'Score'], inplace=True)

In [13]:
questions.head()

Id                                              Title  \
0   80  SQLStatement.execute() - multiple queries in o...   
1   90  Good branching and merging tutorials for Torto...   
2  120                                  ASP.NET Site Maps   
3  180                 Function for creating color wheels   
4  260  Adding scripting functionality to .NET applica...   

                                                Body  
0  <p>I've written a database generation script i...  
1  <p>Are there any really good tutorials explain...  
2  <p>Has anyone got experience creating <strong>...  
3  <p>This is something I've pseudo-solved many t...  
4  <p>I have a little game written in C#. It uses...

In [14]:
top10_tags = list(tags.Tag.value_counts()[:10].index)
print(top10_tags)
new_tags=tags[tags.Tag.isin(top10_tags)]
print(new_tags.shape)
new_tags.reset_index(drop=True, inplace=True)
new_tags.head()

['javascript', 'java', 'c#', 'php', 'android', 'jquery', 'python', 'html', 'c++', 'ios']
(826739, 2)


Id  Tag
0   260   c#
1   330  c++
2   650   c#
3   930   c#
4  1010   c#

In [15]:
questions.isnull().sum()

Id       0
Title    0
Body     0
dtype: int64

In [16]:
new_tags.isnull().sum()

Id     0
Tag    0
dtype: int64

In [17]:
df=questions.merge(new_tags,on='Id')
print(df.shape)
df.head()

(56008, 4)


Id                                              Title  \
0   260  Adding scripting functionality to .NET applica...   
1   330          Should I use nested classes in this case?   
2   650                Automatically update version number   
3   930  How do I connect to a database and loop over a...   
4  1010  How to get the value of built, encoded ViewState?   

                                                Body  Tag  
0  <p>I have a little game written in C#. It uses...   c#  
1  <p>I am working on a collection of classes use...  c++  
2  <p>I would like the version property of my app...   c#  
3  <p>What's the simplest way to connect and quer...   c#  
4  <p>I need to grab the base64-encoded represent...   c#

In [18]:
df.isnull().sum()

Id       0
Title    0
Body     0
Tag      0
dtype: int64

In [19]:
df.drop(columns=['Id'],inplace=True)

In [20]:
X=df.drop(['Tag'],axis=1)
y=df['Tag']

In [21]:
from sklearn.preprocessing import LabelEncoder
LE=LabelEncoder()
y=LE.fit_transform(y)

In [104]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2)

In [105]:
print(X_train.shape, X_test.shape)
print(y_train.shape, y_test.shape)

(44806, 2) (11202, 2)
(44806,) (11202,)


### TEXT PREPROCESSING

In [ ]:
from bs4 import BeautifulSoup

# Converting html to text
X_train['Body']=X_train['Body'].apply(lambda x: BeautifulSoup(x).get_text())
X_test['Body']=X_test['Body'].apply(lambda x: BeautifulSoup(x).get_text())
X_train['Title']=X_train['Title'].apply(lambda x: BeautifulSoup(x).get_text())
X_test['Title']=X_test['Title'].apply(lambda x: BeautifulSoup(x).get_text())

In [ ]:
#To Remove Punctuations
X_train.replace('[^a-zA-Z]',' ', regex=True, inplace=True)
X_test.replace('[^a-zA-Z]',' ', regex=True, inplace=True)

In [ ]:
##Converting to lower case characters
for index in X_train.columns:
  X_train[index] = X_train[index].str.lower()

for index in X_test.columns:
  X_test[index] = X_test[index].str.lower()

In [109]:
#To Remove multiple blank spaces
X_train = X_train.replace('\s+', ' ', regex=True)
X_test = X_test.replace('\s+', ' ', regex=True)

In [110]:
X_train

Title  \
24768  pinch zoom viewport specification works on iph...   
48956  what does the dot operator before the generic ...   
26044  arith overflows in java why there are no excep...   
44314  python log displayed twice when logging module...   
8170   easily position an element over another elemen...   
...                                                  ...   
52801  comparing first element of the consecutive lis...   
4287                                 hashcode uniqueness   
26921  java oracle exception maximum number of expres...   
24821  c virtual functions that need to call the same...   
16658  linq the type of one of the expressions in the...   

                                                    Body  
24768  i made a wordpress site a while back from a sa...  
48956  i saw this code today immutablemap class exten...  
26044  i know this question has been partly answered ...  
44314  context python two files in the same folder fi...  
8170   i need to position a div over an image with jq...  
...                                                  ...  
52801  i have a list of tuples each containing two el...  
4287   is it possible for two instances of object to ...  
26921  i am passing a list of strings to my query sql...  
24821  i have a base class and classes that derive fr...  
16658  i have a complex linq to sql to query which jo...  

[44806 rows x 2 columns]

In [111]:
import nltk
nltk.download('stopwords')
from nltk.stem.snowball import SnowballStemmer
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to C:\Users\R.
[nltk_data]     Harini\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [112]:
type(X_train)

pandas.core.frame.DataFrame

In [113]:
#Removing stopwords
stop_words=set(stopwords.words('english')) 
X_train['Body'] = X_train['Body'].apply(lambda x: ' '.join(term for term in x.split() if term not in stop_words))
X_train['Title'] = X_train['Title'].apply(lambda x: ' '.join(term for term in x.split() if term not in stop_words))

X_test['Body'] = X_test['Body'].apply(lambda x: ' '.join(term for term in x.split() if term not in stop_words))
X_test['Title'] = X_test['Title'].apply(lambda x: ' '.join(term for term in x.split() if term not in stop_words))

In [114]:
print(stop_words)

{'re', "you'll", 'haven', "couldn't", "that'll", 'yours', 'didn', 'a', 'isn', 'aren', 'more', 'over', 'what', 'up', 'i', 'couldn', 'or', "isn't", 'from', 'having', "she's", 'now', 'under', 'won', "mightn't", 'you', 'other', 'can', 'each', 't', 'he', "hadn't", 'theirs', "doesn't", 'ain', 's', 'with', 'should', "you're", 'its', 'of', 'does', 'nor', "you'd", 'again', 'ours', 'will', "haven't", "don't", 'just', 'ma', 'yourself', 'below', "wasn't", 'but', 'as', 'this', 'until', 'do', 'about', 'mustn', 'were', 'where', 'doesn', 'doing', 'if', 'her', 'during', 'because', 'hasn', 'any', 'off', 'are', 'being', 'before', "didn't", 'who', "wouldn't", "won't", 'for', 'those', 'whom', 'through', 'me', 'their', 'to', "hasn't", 'why', "shan't", 'wasn', "shouldn't", 'that', 'the', 'same', "should've", 'your', 'they', 'here', 'd', 'into', 'don', 'weren', 'so', 'most', "needn't", 'o', 'and', 'had', 'been', 'myself', 'ourselves', 'll', 'yourselves', 'mightn', 'be', 'she', 'wouldn', 'him', 'above', 'has',

In [115]:
#Stemming
stemmer = nltk.SnowballStemmer(language='english')

X_train['Body'] = X_train['Body'].apply(lambda x:' '.join(stemmer.stem(term) for term in x.split()))
X_train['Title'] = X_train['Title'].apply(lambda x:' '.join(stemmer.stem(term) for term in x.split()))

X_test['Body'] = X_test['Body'].apply(lambda x:' '.join(stemmer.stem(term) for term in x.split()))
X_test['Title'] = X_test['Title'].apply(lambda x:' '.join(stemmer.stem(term) for term in x.split()))

In [132]:
train_lines = []
for row in range(0,X_train.shape[0]):
    train_lines.append(' '.join(str(x) for x in X_train.iloc[row,:]))

test_lines = []
for row in range(0,X_test.shape[0]):
    test_lines.append(' '.join(str(x) for x in X_test.iloc[row,:]))

In [117]:
train_lines

['pinch zoom viewport specif work iphon android made wordpress site back sail compani xssail com current view port set meta name viewport content user scalabl yes width devic width initi scale maximum scale site work fine comput pinch zoom fine iphon ipad visit site kindl cannot pinch zoom like site idea thank',
 'dot oper generic paramet mean saw code today immutablemap class extend clientcommand commandprocessorinterfac immutablemap immutablemap class extend clientcommand commandprocessorinterfac syntax mean immutablemap class knew generic right class name differ immutablemap class immutablemap class',
 'arith overflow java except run time compil warn know question part answer explain happen arithmet overflow elsewher explain check overflow java code e prevent programm ask except run time compil warn afaik answer bruce eckel book think java st ed chapter littl java program overflow java surpris java let overflow public class overflow public static void main string arg int big x fffff

In [118]:
test_lines

['aspx return imag output cach abl greet creat apsx web form return remot imag base suppli paramet use like img src imagegett aspx param abc param imagegett aspx markup code look similar outputcach durat varybyparam locat serverandcli page languag c autoeventwireup fals enablesessionst fals codebehind imagegett aspx cs inherit acm helper imagegett code call imagegett aspx page load method byte data null dictionari string string file locat getimageloc param param tri data new webclient downloaddata file locat main catch webexcept wex tri data new webclient downloaddata file locat backup catch except e throw respons contenttyp imag jpeg respons outputstream write data data length respons end test appear cach possibl output cach resort write cach store byte array depend queri string paramt',
 'object map librari json nsobject tri build parser objectmapp build object c object json consum rest servic took inspir restkit entiti hold decod list tell mapper json key goe object like objectent i

In [119]:
len(train_lines), len(test_lines)

(44806, 11202)

In [120]:
from sklearn.feature_extraction.text import CountVectorizer

countvector=CountVectorizer()
X_train_cv=countvector.fit_transform(train_lines)
X_test_cv=countvector.transform(test_lines)

In [121]:
from sklearn.feature_extraction.text import TfidfVectorizer, TfidfTransformer

tfidfvector=TfidfTransformer()
X_train_tf=tfidfvector.fit_transform(X_train_cv)
X_test_tf=tfidfvector.fit_transform(X_test_cv)

### Logistic Regression

In [122]:
from sklearn.linear_model import LogisticRegression
from sklearn.multiclass import OneVsRestClassifier 

lr=LogisticRegression()
clf=OneVsRestClassifier(lr)

In [123]:
clf.fit(X_train_tf, y_train)

OneVsRestClassifier(estimator=LogisticRegression(C=1.0, class_weight=None,
                                                 dual=False, fit_intercept=True,
                                                 intercept_scaling=1,
                                                 l1_ratio=None, max_iter=100,
                                                 multi_class='auto',
                                                 n_jobs=None, penalty='l2',
                                                 random_state=None,
                                                 solver='lbfgs', tol=0.0001,
                                                 verbose=0, warm_start=False),
                    n_jobs=None)

In [124]:
y_pred=clf.predict(X_test_tf)
y_pred

array([1, 4, 5, ..., 1, 4, 3])

In [125]:
y_pred[y_pred==9]

array([9, 9, 9, ..., 9, 9, 9])

In [126]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
print(confusion_matrix(y_test, y_pred))

[[1102   22    7    7   22   96   10    1    1    1]
 [   9 1565   61    7    6   79   20    5   13   16]
 [   4  119  914    3    4   44   11    0    3   18]
 [  10   19    4  268    2   10  129   56   26   10]
 [  23   35   16    5  589   10   10    1    5    3]
 [ 115   89   26    6    2 1483   15    2    9    8]
 [  13   55    9  103   12   19  947  170   36   13]
 [   5   17    0   59    3    5  278  351   21    2]
 [   2   46    6   16    2   14   28   11  656    7]
 [   5   35   10    6    0   19    6    0    6 1053]]


In [127]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.87      0.86      1269
           1       0.78      0.88      0.83      1781
           2       0.87      0.82      0.84      1120
           3       0.56      0.50      0.53       534
           4       0.92      0.85      0.88       697
           5       0.83      0.85      0.84      1755
           6       0.65      0.69      0.67      1377
           7       0.59      0.47      0.52       741
           8       0.85      0.83      0.84       788
           9       0.93      0.92      0.93      1140

    accuracy                           0.80     11202
   macro avg       0.78      0.77      0.77     11202
weighted avg       0.80      0.80      0.80     11202



In [128]:
print(accuracy_score(y_test, y_pred))

0.7970005356186395


### SVM

In [44]:
from sklearn.svm import SVC
clf=SVC(kernel='linear').fit(X_train_tf, y_train)

In [45]:
preds=clf.predict(X_test_tf)
preds

array([2, 2, 6, ..., 5, 8, 9])

In [46]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
print(confusion_matrix(y_test, preds))

[[1167   29    6    6   14  110   17    0    0    2]
 [  16 1560   82   23   10   60   37   11   13    9]
 [  11  112  941    2    9   41    8    0    5   11]
 [   3   10    6  292    2    4  142   59   24    2]
 [  13   29    6    3  563   10   19    0    7    2]
 [ 139   99   32   13    0 1374   23    3   13    8]
 [  21   60    4   82   12   15  895  266   32   10]
 [   2   13    1   57    4    3  251  349   14    1]
 [   6   45    5   15    2   13   26   12  640    3]
 [   1   53   22   11    1   16   23    1    9  994]]


In [47]:
print(classification_report(y_test, preds))

              precision    recall  f1-score   support

           0       0.85      0.86      0.85      1351
           1       0.78      0.86      0.81      1821
           2       0.85      0.83      0.84      1140
           3       0.58      0.54      0.56       544
           4       0.91      0.86      0.89       652
           5       0.83      0.81      0.82      1704
           6       0.62      0.64      0.63      1397
           7       0.50      0.50      0.50       695
           8       0.85      0.83      0.84       767
           9       0.95      0.88      0.91      1131

    accuracy                           0.78     11202
   macro avg       0.77      0.76      0.77     11202
weighted avg       0.79      0.78      0.78     11202



In [48]:
print(accuracy_score(y_test, preds))

0.7833422603106588


### NAIVE BAYES

In [134]:
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB().fit(X_train_tf, y_train)

In [135]:
preds = clf.predict(X_test_tf)
preds

array([1, 1, 5, ..., 1, 4, 6])

In [136]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
print(confusion_matrix(y_test, preds))

[[ 930   44    5    0    0  249   41    0    0    0]
 [   2 1616   11    0    0  110   38    0    0    4]
 [   6  251  667    0    0  161   14    0    1   20]
 [   4   24    0   19    0   18  460    0    4    5]
 [  88  160   10    0  252   99   87    0    1    0]
 [  84  109   11    0    0 1522   27    0    0    2]
 [   5  120    1    1    0   45 1197    1    2    5]
 [   3   28    0    1    0    5  702    0    2    0]
 [   5  241    1    0    0  193  202    0  138    8]
 [   8  183    8    0    0  174   55    0    0  712]]


In [137]:
print(classification_report(y_test, preds))

              precision    recall  f1-score   support

           0       0.82      0.73      0.77      1269
           1       0.58      0.91      0.71      1781
           2       0.93      0.60      0.73      1120
           3       0.90      0.04      0.07       534
           4       1.00      0.36      0.53       697
           5       0.59      0.87      0.70      1755
           6       0.42      0.87      0.57      1377
           7       0.00      0.00      0.00       741
           8       0.93      0.18      0.29       788
           9       0.94      0.62      0.75      1140

    accuracy                           0.63     11202
   macro avg       0.71      0.52      0.51     11202
weighted avg       0.69      0.63      0.59     11202



In [138]:
print(accuracy_score(y_test, preds))

0.6296197107659347


#### accuracy of Naive Bayes              62.96%
#### accuracy of SVM                            78.33%
#### accuracy of Logistic Regression 79.70%